MCOL-498 Optionally disable disk space preallocation.

drrtuy opened the pull request:
Plz look into JIRA's issue for details.

drrtuy committed:
MCOL-498. Segment files extension uses fallocate() now to optimize load put on SSD disks.

drrtuy committed:
MCOL-498. Add the knob to disable segment|dict file preallocation. Dict files extension uses fallocate() if possible.

drrtuy committed:
MCOL-498: Fill up the block with NULLs when CS touches for the first time it with INSERT..VALUES.

drrtuy committed:
MCOL-498: Fill up the last used block with empty values, whilst doing bulk insertion with uncompressed files.

drrtuy committed:
MCOL-498 Fill up next block with empty values if insert values up to the block boundary.

drrtuy committed:
MCOL-498 Support for non dict compressed columns.

drrtuy committed:
MCOL-498 Passed test100.

drrtuy committed:
MCOL-498 Add DBRootX.PreallocSpace setting in the XML. Dict files extents now contain a correct number of blocks available.

pleblanc1976 commented on utils/idbdatafile/CMakeLists.txt:
```diff
@@ -16,7 +16,7 @@ set(idbdatafile_LIB_SRCS
 add_library(idbdatafile SHARED ${idbdatafile_LIB_SRCS})
 
-target_link_libraries(idbdatafile ${NETSNMP_LIBRARIES})
+target_link_libraries(idbdatafile ${NETSNMP_LIBRARIES} ${ENGINE_OAM_LIBS})
```
note to self.  I don't see a change yet that requires the new dependency.

pleblanc1976 commented on utils/idbdatafile/IDBPolicy.cpp:
```diff
@@ -224,9 +228,73 @@ void IDBPolicy::configIDBPolicy()
+    // there is no DBRootX sections in XML.
+    if (pt == moduledbrootlist.end())
+    {
+        return;
```
Would this indicate no dbroots are assigned to this PM?

drrtuy replied:
Yes. It would. Is this relevant in this case?

pleblanc1976 replied:
No, it's just for my own understanding of the code.

pleblanc1976 commented on writeengine/shared/we_fileop.cpp:
```diff
@@ -1096,69 +1100,100 @@ int FileOp::initColumnExtent(
+        // for the dbroot
+        if ( bOptExtension )
+        {
+            bOptExtension = (idbdatafile::IDBPolicy::PreallocSpace(dbRoot))
```
since bOptExtension is true here, this reduces to bOptExtension = ..PreallocSpace(dbRoot)

drrtuy replied:
bOptExtension is set when CS increases an extent only and not for abbreviated extent so I prefered conditional jump to a function call.

pleblanc1976 commented on writeengine/bulk/we_colbuf.cpp:
```diff
@@ -101,15 +102,33 @@ void ColumnBuffer::resizeAndCopy(int newSize, int startOffset, int endOffset)
+        nitems = fFile->write(fBuffer + startOffset, writeSize) / writeSize;
 
     if (nitems != 1)
         return ERR_FILE_WRITE;
```
need to dealloc newBuf before returning

drrtuy replied:
Good catch.

pleblanc1976 commented on writeengine/bulk/we_colbuf.cpp:
```diff
@@ -101,15 +102,33 @@ void ColumnBuffer::resizeAndCopy(int newSize, int startOffset, int endOffset)
 // Write data stored up in the output buffer to the segment column file.
 //------------------------------------------------------------------------------
-int ColumnBuffer::writeToFile(int startOffset, int writeSize)
+int ColumnBuffer::writeToFile(int startOffset, int writeSize, bool fillUpWNulls)
```
Optional.  fillUpWNulls should probably be called fillUpWEmpty instead.

drrtuy replied:
Strongly agree on that.

pleblanc1976 commented on writeengine/bulk/we_colbufcompressed.cpp:
```diff
@@ -242,12 +247,14 @@ int ColumnBufferCompressed::writeToFile(int startOffset, int writeSize)
             fLog->logMsg( oss.str(), MSGLVL_INFO2 );
         }
 
         memcpy(bufOffset, (fBuffer + startOffset), writeSize);
```
I'm a little unclear how this is supposed to work.  It might make more sense later.  With my current understanding, the empty values should be copied to bufOffset here...?

pleblanc1976 replied:
A little more reading makes me think the buffer is initialized with empty values, so that would be unnecessary.  Please confirm that behavior

drrtuy replied:
It is not only about the buffer initialization but about how many bytes was written.
This additional bytes must make singular INSERTs happy but I am looking into it in details now.

drrtuy replied:
This change is unnecessary indeed. Nice catch.

pleblanc1976 replied:
Groovy

pleblanc1976 commented on writeengine/wrapper/we_colop.cpp:
```diff
@@ -1533,12 +1542,16 @@ int ColumnOp::writeRow(Column& curCol, uint64_t totalRow, const RID* rowIdArray,
     uint64_t  emptyVal;
     int rc = NO_ERROR;
+    bool fillUpWEmptyVals = false;
+    bool fistRowInBlock = false;
```
a typo worth fixing

drrtuy replied:
Fixed.

pleblanc1976 replied:
Sorry, I meant 'fistRowInBlock' -> 'firstRowInBlock'.  :)

drrtuy replied:
I didn't miss the typo for the second time :)

pleblanc1976 reviewed (changes requested):
Some of this was pretty hard to understand without reading through all of the writeengine code.  The harder it is to read the code, the more we have to rely on testing.  A couple comments...
1) Replace all of the variable names & comments that suggest we're padding with NULL values rather than empty values.
2) Add a component test or two.  Ex: a script to import a column, big & small, with preallocation on and off, then verify the expected result from dumpcol or coldump (whatever it's called).

drrtuy commented:
As you suggested I will add a test for this and comment most of the changes to make it more readable.

drrtuy committed:
MCOL-498 Changes made according with review suggestions.
    Add more comments.
    Changed return value for HDFS'es fallocate.
    Removed unnecessary code in ColumnBufferCompressed::writeToFile
    Replaced Nulls with Empties in variable names.
